src/rust/catnap/win/overlapped.rs

```diff
@@ -0,0 +1,774 @@
+                                    Ok(Rc::clone(Pin::get_mut(state)))
+                                }
+                            },
+
+                            Err(fail) => Err(fail),
+                        }
+                    },
+                )
+            }
+            .await?;
+
+            let message: &str = std::str::from_utf8(buffer.as_slice())
+                .map_err(|_| Fail::new(libc::EINVAL, "utf8 conversion failed"))?;
+            if message != MESSAGE {
+                let err_msg: String = format!("expected \"{}\", got \"{}\"", MESSAGE, message);
+                Err(Fail::new(libc::EINVAL, err_msg.as_str()))
+            } else {
+                Ok(())
+            }
+        });
+
+        let mut scheduler: Scheduler = Scheduler::default();
+        let server_handle: TaskHandle = scheduler
+            .insert(TaskWithResult::<Result<(), Fail>>::new("server".into(), server))
+            .unwrap();
+
+        let get_server_result = |scheduler: &mut Scheduler| {
+            TaskWithResult::<Result<(), Fail>>::from(scheduler.remove(&server_handle).unwrap().as_any())
+                .get_result()
+                .unwrap()
+        };
+
+        let mut wait_for_state = |scheduler: &mut Scheduler, state| -> Result<(), Fail> {
+            while server_state_view.load(Ordering::Relaxed) < state {
+                iocp.get_mut().process_events()?;
+                scheduler.poll();
+                if server_handle.has_completed() {
+                    return Err(get_server_result(scheduler).unwrap_err());
+                }
+            }
+            Ok(())
+        };
+
+        wait_for_state(&mut scheduler, 1)?;
+
+        let client_handle: SafeHandle = SafeHandle(unsafe {
+            CreateFileA(
+                PIPE_NAME,
+                GENERIC_WRITE.0,
+                FILE_SHARE_NONE,
+                None,
+                OPEN_EXISTING,
+                FILE_FLAGS_AND_ATTRIBUTES::default(),
+                HANDLE(0),
+            )
+        }?);
+
+        wait_for_state(&mut scheduler, 2)?;
+
+        let mut bytes_written: u32 = 0;
+        unsafe {
+            WriteFile(
+                client_handle.0,
+                Some(MESSAGE.as_bytes()),
+                Some(&mut bytes_written),
+                None,
+            )?;
+        }
+
+        std::mem::drop(client_handle);
+
+        while !server_handle.has_completed() {
+            iocp.get_mut().process_events()?;
+            scheduler.poll();
+        }
+
+        get_server_result(&mut scheduler)?;
+
+        Ok(())
+    }
+
+    /// Test I/O cancellation.
+    #[test]
+
+    fn test_cancel_io() -> Result<()> {
+        const PIPE_NAME: PCSTR = s!(r"\\.\pipe\demikernel-test-cancel-pipe");
+        const BUFFER_SIZE: u32 = 128;
+        const COMPLETION_KEY: usize = 0xFEEDF00D;
+        let server_pipe: SafeHandle = SafeHandle(unsafe {
+            CreateNamedPipeA(
+                PIPE_NAME,
```

Comment:
Is some cleanup required before successive runs of this test in case of failures/abrupt test termination?